tensorflow_recommenders_addons/dynamic_embedding/core/kernels/lookup_impl/lookup_table_op_cpu.h

```diff
@@ -110,6 +110,96 @@ struct HybridHash<int32> {
   }
 };
 
+template <typename T>
+class HostFileBuffer {
+ public:
+  enum MODE { READ = 0, WRITE = 1 };
+
+  HostFileBuffer(OpKernelContext* ctx, const string path, size_t size,
+                 MODE mode)
+      : ctx_(ctx), filepath_(path), capacity_(size), mode_(mode) {
+    offset_ = 0;
+    buf_ = (T*)malloc(capacity_ * sizeof(T));
+    OP_REQUIRES(ctx, buf_ != nullptr,
+                errors::ResourceExhausted("Failed to open file ", filepath_));
+    if (mode_ == MODE::READ) {
+      fp_ = fopen(filepath_.c_str(), "rb");
+    } else if (mode == MODE::WRITE) {
+      fp_ = fopen(filepath_.c_str(), "wb");
+    } else {
+      fp_ = nullptr;
+    }
+    OP_REQUIRES(ctx, fp_ != nullptr,
+                errors::ResourceExhausted("Failed to open file ", filepath_));
+  }
+
+  void Close() {
+    if (buf_) {
+      free(buf_);
+      buf_ = nullptr;
+    }
+    if (fp_) {
+      fclose(fp_);
+      fp_ = nullptr;
+    }
+  }
+
+  ~HostFileBuffer() { Close(); }
+
+  void Put(const T value) {
```

Comment:
May I ask if T is `tstring`, will it work? I remember that sometimes the key is tstring.

Reply:
Haven't used string key type. I think it can be solved by abstracting the host buffer class and template specialization.